### DISCUSSION TITLE
Is there a best practice on how to declare/manage queryKey?

### DISCUSSION THREAD

**Discussion by hyungjunk** [TanStack/query#3362] [Q&A]
A project gets bigger and bigger and there be `useQuery` or `useMutation` codes in the project.

Therefore, it seems worthy setting a convention on how to create queryKey.

Would there be recommended or best practice how to manage it?

The practice suggested here [blog posting](https://tkdodo.eu/blog/effective-react-query-keys) doesn't seem the best for me because potentially the queryKey could be duplicated even if the parameter is also included to the queryKey.

**Comment by TkDodo** [comment]
> The practice suggested here [blog posting](https://tkdodo.eu/blog/effective-react-query-keys) doesn't seem the best for me because potentially the queryKey could be duplicated even if the parameter is also included to the queryKey.

can you elaborate on the duplication issue ?

**Reply by hyungjunk** [reply]
@TkDodo Sure, I believe that it wouldn't happen frequently and if it would, it would probably be because of human error. Say that I am using `[userName, ${userId}]` as a queryKey. What if someone just didn't know that this key is already occupied and he declares another query function with the same key? or the key is dynamically created `[getUserName(), ${userId}]` but eventually the same key is used? To prevent this, I would like to make a convention about the best practice to make a queryKey without worrying about duplication.

**Reply by TkDodo** [reply]
FWIW, we organize our project into `features`. Each `feature` has its own queryKeys that start with the name of the feature to avoid clashes, like:

```
- src
  - features
    - profile
    - dashboard
    - todos
```

in each of those, we have a `queries.ts` file that has a `queryKeyFactory`, scoped to the feature. No clashes here :)
